### PR TITLE
Update Account Schema

### LINKED ISSUE #52: Update Account Schema
In order to support Company and Admin accounts, the Account schema should be updated to contain the following fields:
- `email`
- `password`
- `isAdmin` : Boolean
- `company` : Ref to Company Model

Additionally, if the Account is admin (`isAdmin === true`), `company` must be null. If `company` is non-null, `isAdmin` must be false.

Furthermore, the following middlewares must be created:
- `isAdmin` - the logged in user is an admin
- `belongsToCompany(companyId)` - the logged in user is a representative of the company

### PR DESCRIPTION
**Pull request by GuilhermeJSilva**
Closes #52 

* Added `isAdmin` and `company` fields to Account Schema
* Force only one of the fields `isAdmin` and `company` to be required
* Rename Admin Token to God Token
* Make register endpoint force the created user to be an admin
* Only the god user can create admins

### CODE REVIEW

File: src/api/middleware/validators/auth.js [hunk 1]
```diff
@@ -3,19 +3,20 @@ const { body } = require("express-validator");
 const { useExpressValidators } = require("../errorHandler");
 const Account = require("../../../models/Account");
 
-const checkDuplicateUsername = async (username) => {
-    const acc = await Account.findOne({ username }).exec();
+const checkDuplicateEmail = async (email) => {
```
Comment (imnotteixeira) [inline src/api/middleware/validators/auth.js]:
```suggestion
const checkDuplicatedEmail = async (email) => {
```

File: src/api/middleware/validators/auth.js [hunk 2]
```diff
@@ -3,19 +3,20 @@ const { body } = require("express-validator");
...
-    body("username", "Invalid username")
-        .exists().withMessage("Username is required").bail()
-        .isString().withMessage("Username must be a String")
+    body("email", "Invalid email")
+        .exists().withMessage("Email is required").bail()
+        .isString().withMessage("Email must be a String")
+        .isEmail().withMessage("Email must be valid")
         .bail()
-        .custom(checkDuplicateUsername)
+        .custom(checkDuplicateEmail)
```
Comment (imnotteixeira) [inline src/api/middleware/validators/auth.js]:
```suggestion
        .custom(checkDuplicatedEmail)
```

File: src/api/middleware/validators/auth.js [hunk 3]
```diff
@@ -3,19 +3,20 @@ const { body } = require("express-validator");
...
     }
 };
 
 const register = useExpressValidators([
-    body("username", "Invalid username")
-        .exists().withMessage("Username is required").bail()
-        .isString().withMessage("Username must be a String")
+    body("email", "Invalid email")
+        .exists().withMessage("Email is required").bail()
+        .isString().withMessage("Email must be a String")
```
Comment (imnotteixeira) [inline src/api/middleware/validators/auth.js]: Do we need to check "is String" when we are checking "is Email"? It's kind of redundant

File: src/config/passport.js [hunk 4]
```diff
@@ -4,21 +4,24 @@ const LocalStrategy = require("passport-local").Strategy;
...
+    usernameField: "email",
+    passwordField: "password",
+},
+(email, password, done) => {
+    Account.findOne({ email: email }, (err, user) => {
+        if (err) {
+            return done(err);
+        }
+        if (!user) {
+            return done(null, false, { message: "Incorrect email." });
```
Comment (imnotteixeira) [inline src/config/passport.js]: Shouldn't this return a message like "Incorrect email or password."? I understand that in this scope, it makes sense that the errors are separated, but I believe this message is being served as an error, therefore it should be generalized

File: src/config/passport.js [hunk 5]
```diff
@@ -4,21 +4,24 @@ const LocalStrategy = require("passport-local").Strategy;
...
+        }
+        if (!user.validatePassword(password)) {
+            return done(null, false, { message: "Incorrect password." });
```
Comment (imnotteixeira) [inline src/config/passport.js]: Same as above ("Incorrect email or password.")

File: src/api/middleware/validators/auth.js [hunk 6]
```diff
@@ -3,19 +3,19 @@ const { body } = require("express-validator");
...
     }
 };
 
 const register = useExpressValidators([
-    body("username", "Invalid username")
-        .exists().withMessage("Username is required").bail()
-        .isString().withMessage("Username must be a String")
+    body("email", "Invalid email")
+        .exists().withMessage("Email is required").bail()
+        .normalizeEmail().isEmail().withMessage("Email must be valid")
```
Comment (miguelpduarte) [inline src/api/middleware/validators/auth.js]: This should be using the validator reasons like in the other endpoint validators. The messages don't matter as their content will be decided in the Front-End. As such, we are using enums in the validators.

File: src/api/middleware/validators/auth.js [hunk 7]
```diff
@@ -3,19 +3,19 @@ const { body } = require("express-validator");
 const { useExpressValidators } = require("../errorHandler");
 const Account = require("../../../models/Account");
 
-const checkDuplicateUsername = async (username) => {
-    const acc = await Account.findOne({ username }).exec();
+const checkDuplicatedEmail = async (email) => {
+    const acc = await Account.findOne({ email }).exec();
     if (acc) {
-        throw new Error("Username already exists");
+        throw new Error("Email already exists");
```
Comment (miguelpduarte) [inline src/api/middleware/validators/auth.js]: I'm not entirely sure what message is being sent in other custom validators, but we should probably also use the validator reasons enum here (and elsewhere).

File: src/api/middleware/validators/auth.js [hunk 8]
```diff
@@ -25,4 +25,15 @@ const register = useExpressValidators([
...
-module.exports = { register };
+const login =  useExpressValidators([
+    body("email", "Invalid email")
+        .exists().withMessage("Email is required").bail()
+        .normalizeEmail().isEmail().withMessage("Email must be valid")
+        .trim(),
+
+    body("password", "Invalid password")
+        .exists().withMessage("Password is required").bail()
+        .isString().withMessage("Password must be a String"),
```
Comment (miguelpduarte) [inline src/api/middleware/validators/auth.js]: The messages should be changed, see above.

File: src/config/passport.js [hunk 9]
```diff
@@ -4,21 +4,21 @@ const LocalStrategy = require("passport-local").Strategy;
...
-            }
-            return done(null, user);
-        });
-    }
+passport.use(new LocalStrategy({
+    usernameField: "email",
+    passwordField: "password",
+},
+(email, password, done) => {
+    Account.findOne({ email: email }, (err, user) => {
```
Comment (miguelpduarte) [inline src/config/passport.js]:
```suggestion
    Account.findOne({ email }, (err, user) => {
```

File: src/models/Account.js [hunk 10]
```diff
@@ -3,8 +3,39 @@ const bcrypt = require("bcrypt");
...
+    isAdmin: {
+        type: Boolean,
+        required: function() {
+            return !this.companyRef;
+        },
+        validate: {
+            validator: function(isAdmin) {
+                return isAdmin && !this.company;
+            },
+            message: "A user can not be an admin and a company representative",
```
Comment (miguelpduarte) [inline src/models/Account.js]:
```suggestion
            message: "A user cannot be an admin and a company representative at the same time",
```

File: src/models/Account.js [hunk 11]
```diff
@@ -3,8 +3,39 @@ const bcrypt = require("bcrypt");
...
+    company: {
+        type: Schema.Types.ObjectId, ref: "Company",
+        required: function() {
+            return !this.isAdmin;
+        },
+        validate: {
+            validator: function(companyRef) {
+                return !!companyRef && !this.isAdmin;
+            },
+            message: "A user can not be an admin and a company representative",
```
Comment (miguelpduarte) [inline src/models/Account.js]: Same as above

File: test/end-to-end/auth.js [hunk 12]
```diff
@@ -1,27 +1,33 @@
...
 
                 expect(res.status).toBe(422);
                 expect(res.body).toHaveProperty("error_code", ErrorTypes.VALIDATION_ERROR);
                 expect(res.body).toHaveProperty("errors");
                 expect(res.body.errors).toContainEqual({
                     "location": "body",
-                    "msg": "Username is required",
-                    "param": "username",
+                    "msg": "Email is required",
+                    "param": "email",
```
Comment (miguelpduarte) [inline test/end-to-end/auth.js]: This (and all other tests that can do so) should be using the abstractions for testing validators that were introduced in the previous PR.

File: test/end-to-end/auth.js [hunk 13]
```diff
@@ -1,27 +1,33 @@
 const { ErrorTypes } = require("../../src/api/middleware/errorHandler");
 const Account = require("../../src/models/Account");
 
+const test_god_token  = "testing_is_cool73";
```
Comment (miguelpduarte) [inline test/end-to-end/auth.js]: Should be using `withGodToken`

**Comment by GuilhermeJSilva** [comment]
>     * Fix custom validator for duplicate email to use a ValidatorReason
> 
>     * Add test for duplicate email responding with validation error
> 
> 
> Other than that, incredible! Good job!

The test for duplicated email was already done, but I added the reason to the ValidationReasons object
